cpp/merkletree/sparse_merkle_tree.cc

```diff
@@ -0,0 +1,216 @@
+#include "cpp/merkletree/sparse_merkle_tree.h"
+
+#include <algorithm>
+#include <stddef.h>
+#include <vector>
+
+#include "merkletree/merkle_tree_math.h"
+#include "util/util.h"
+
+using std::make_pair;
+using std::ostream;
+using std::ostringstream;
+using std::reverse;
+using std::string;
+using std::unique_ptr;
+using std::unordered_map;
+using std::vector;
+
+
+const vector<string>& GetNullHashes(const TreeHasher& hasher) {
+  static unique_ptr<const vector<string>> null_hashes;
+  if (!null_hashes.get()) {
+    vector<string> r{hasher.HashLeaf("")};
+    for (int i(1); i < hasher.DigestSize() * 8; ++i) {
+      r.emplace_back(hasher.HashChildren(r.back(), r.back()));
+    }
+    reverse(r.begin(), r.end());
+    null_hashes.reset(new vector<string>(r.begin(), r.end()));
+  }
+  return *null_hashes.get();
+}
+
+
+SparseMerkleTree::SparseMerkleTree(SerialHasher* hasher)
+    : serial_hasher_(CHECK_NOTNULL(hasher)->Create()),
+      treehasher_(hasher),
+      null_hashes_(GetNullHashes(treehasher_)),
+      root_hash_(null_hashes_[0]) {
+}
+
+
+void SparseMerkleTree::EnsureHaveLevel(size_t level) {
+  if (tree_.size() < (level + 1)) {
+    tree_.emplace_back(unordered_map<IndexType, TreeNode>());
```

Comment:
I think if that ever happened it'd be because of a bug elsewhere, but I'll change it for `resize()` anyway.